Autopilot/PathManager/Inc/waypointManager.hpp

```diff
@@ -0,0 +1,175 @@
+#define HOLD_WAYPOINT 2 // Circle
+
+// Placeholder: 
+typedef struct {
+    /*
+        Contains:
+            long 
+            lat
+            alt
+            heading
+    */    
+} _Gps_Data;
+
+// Stores error codes for the waypoint manager
+enum _WaypointStatus {WAYPOINT_SUCCESS = 0, WAYPOINT_NOT_FOUND, WAYPOINT_PARAMETERS_NOT_DEFIND, UNDEFINED_FALIURE};
+
+// Used in the waypointBufferStatus array to signal which elements are free
+enum _WaypointBufferStatus {FREE = 0, FULL};
+
+//Used to specify the modification type when updating the waypointBuffer array
+enum _WaypointBufferUpdateType {ADD_WAYPOINT = 0, UPDATE_WAYPOINT, INSERT_WAYPOINT, DELETE_WAYPOINT};
+
+
+/*
+* Structure stores information about the waypoints along our path to the destination and back. 
+*/
+typedef struct {
+    struct _WaypointData * next;      // Next waypoint
+    struct _WaypointData * previous;  // Previous waypoint
+    long double latitude;             // Latitude of waypoint
+    long double longitude;            // Longitude of waypoint
+    int altitude;                     // Altitude of waypoint
+    float turnRadius;                 // If turn is commanded, this is the radius of the turn
+    char type;
+    char waypointId;                  // Id of the waypoint
+    char arrayIndex;                  // Position of the waypoint within the waypointBuffer (array)
+} _WaypointData;
+
+/*
+* Structure contains the data that will be returned to the Path Manager state manager. 
+* This data will be used by the PID and coordinated turn engine to determine the commands to be sent to the Attitude Manager.
+*/
+typedef struct {
+    uint16_t desiredHeading;            // Desired heading to stay on path
+    int desiredAltitude;                // Desired altitude at next waypoint
+    long double distanceToNextWaypoint; // Distance to the next waypoint (helps with airspeed PID)
+    _WaypointStatus errorCode;          // Contains error codes
+    bool isDataNew;                     // Notifies PID modules if the data in this structure is new
+    uint32_t timeOfData;                // The time that the data in this structure was collected
+} _WaypointManager_Data;
+
+class WaypointManager {
+    public: 
+        /*
+        * Constructor for this class
+        * 
+        * @param _WaypointData * initialWaypoints -> These waypoints will be used to initialize the waypointBuffer array
+        * @param int numWaypoints -> Number of waypoints being initialized in the waypointBuffer array
+        * @param _WaypointData* currentLocation -> Home base. 
+        */
+        WaypointManager(_WaypointData * initialWaypoints, int numWaypoints, _WaypointData* currentLocation);
+        
+        /*
+        * Updates the _WaypointManager_Data structure with new values
+        * 
+        * @param _Gps_Data currentPosition -> contains the current coordinates, altitude, and heading
+        */
+        void get_next_directions(_Gps_Data currentPosition); 
```

Comment:
So is the idea that every time someone wants to use your module, they at least have to call 

get_next_direction and then GetResult ?

Why are these 2 separate functions ? Could they just be put together into 1 function that takes in input and produces output ?

Reply:
True, I can blend them together!